signature/src/mechanism/mod.rs

```diff
@@ -3,4 +3,76 @@
 // SPDX-License-Identifier: Apache-2.0
 //
 
+//! # Signing schemes
+//! different signatures defination and the top level interfaces.
+//!
+//! ### Design
+//! We use both an enum and a struct for a signing scheme.
+//! * `SignSchemeEnum`: The member inside the enum value is a
+//! struct, which implements the trait `SignScheme`.
+//! * `SignScheme`: A trait containing basic interface for the
+//! sign scheme.
+//!
+//! Why we use an extra enum here is to help divide traits into
+//! object safe and object unsafe traits (like `Deserialize`, `Serialize`).
+//! The enum has all the object unsafe traits, and the trait has
+//! object safe traits. Thus enum can be deserialized from `policy.json`.
+//! Also, by call `inner_ref()` can get the reference of the inner
+//! trait object to do all the trait functions.
+
+use std::collections::HashMap;
+
+use anyhow::*;
+use serde::{Deserialize, Serialize};
+
+use crate::Image;
+
+use self::simple::SimpleParameters;
+
 pub mod simple;
+
+// TODO: Add more signature mechanism.
+//
+// Refer to issue: https://github.com/confidential-containers/image-rs/issues/7
+
+/// Signing schemes enum.
+/// * `SimpleSigning`: Redhat simple signing.
+#[derive(Deserialize, Debug, PartialEq, Serialize)]
+#[serde(tag = "scheme")]
+pub enum SignSchemeEnum {
+    #[serde(rename = "simple")]
+    SimpleSigning(SimpleParameters),
+}
+
+/// The interface of a signing scheme
+pub trait SignScheme {
+    /// Prepare runtime directories for storing signatures, configuretions ,.etc
+    fn prepare_runtime_dirs(&self) -> Result<()>;
+
+    /// Check whether the required resources need to be obtained from KBS.
+    fn resources_check(&self) -> Result<bool>;
+
+    /// Get the resources list. The list contains resources names to be
+    /// got by the kbs, using `get_resource_from_kbs()` in validate.rs
+    fn needed_resources_list_from_kbs(&self) -> Vec<&str>;
+
+    /// Process all the gathered resources from kbs. All the resources
+    /// gathered from kbs due to the return value of `needed_resources_list_from_kbs()`
+    /// will be recorded into a HashMap, mapping `resource name` ->
+    /// `content of Vec<u8>`. The parameter of this function is the
+    /// HashMap.
+    fn process_gathered_resources(&self, resources: HashMap<&str, Vec<u8>>) -> Result<()>;
```

Comment:
The name may cause some confusion, perhaps names like `set_dependency` or some other is better?

Reply:
I gave a doc and figure to show how the functions work. I think it will be much clearer.